io-sim/src/Control/Monad/IOSim/Internal.hs

```diff
@@ -230,15 +237,100 @@ schedule !thread@Thread{
         let thread' = thread { threadControl = ThreadControl (k x) ctl' }
         schedule thread' simstate
 
+      TimeoutFrame _tid tmid k ctl' ->
+        -- It could happen that the timeout action finished at the same time
+        -- as the timeout expired, this will be a race condition. That's why
+        -- we have the locks to solve this. If at this moment the lock is
+        -- 'Locked', then it means that this thread still hasn't received the
+        -- 'TimeoutException', so we need to kill the thread that is responsible
+        -- for doing that and run our continuation successfully; if it is
+        -- 'NotLocked' then it means we can lock it and carry on with the
+        -- success case'.
+        case Map.lookup tmid locks of
+          Nothing            -> error "TimeoutFrame: impossible happened!"
+          Just (Locked etid) -> do
+                -- Remove the timer from the queue
+            let timers' = PSQ.delete tmid timers
+                -- Remove the lock for this particular timeout
+                locks'  = Map.delete tmid locks
+                -- Kill the exception throwing thread and carry on the
+                -- continuation
+                thread' = thread { threadControl =
+                                      ThreadControl (ThrowTo (toException ThreadKilled)
+                                                             etid
+                                                             (k (Just x)))
+                                                    ctl'
+                                  }
```

Comment:
So there's still a residual race here isn't there? We're now leaving our `TimeoutFrame` and we're throwing an exception to the other thread that's trying to kill us. So what can happen?

We've got two threads, each of which has as their next action to kill the other one. Due to the semantics of `throwTo` only one of these can succeed, not both (`throwTo` is in fact synchronous, it blocks until the exception is delivered).

So if this thread succeeds first, it will kill the other one, and proceed happily.

If the other thread succeeds first, it will throw a `TimeoutException` to us, but we're now outside of our `TimeoutFrame`! So things will not work as expected.

Thus I conclude that this locking scheme does not do what is intended!

Take another look at the locking scheme that GHC uses, it's not the same as what we're doing here. In the GHC timeout code, the lock is not taken when the interrupting thread is forkIO'd, but is only taken by that forkIO'd thread when it's about to kill the action thread (I'm not sure that's important, but does reduce the time interval). Furthermore, although it does look like a similar race can happen there: note that the cleanup handler uses `uninterruptibleMask_` to prevent itself being killed by the other thread. That does resolve the race.

Reply:
GHC implementation is more subtle because it uses `registerTimeout`, `handleTimeout` and `cleanupTimeout` it also uses `tryPutMVar` and it makes it hard to translate into IOSim code, but here is my understanding. `handleTimeout` runs when the timeout expired, so it will check the lock: if it is `NotLocked` (it means that the running thread still hasn't finished, so it will fork a new thread. This new thread objective is to stop/kill the running thread since the timeout expired, **_however_** right before `throwingTo` it will try to acquire the lock, meaning it will only send the exception after immediately being able to acquire the lock. If the lock is `Locked` in `handleTimeout` it means that the running Thread won the race. Notice that that the `handleTimeout` code runs unmasked.

`cleanupTimeout` if the running thread finished and this can race with the timeout firing at the same time. `cleanupTimeout` is going to try acquire the lock, if it succeeds it just cleans the timer, otherwise it will attempt to kill the thread that is trying to kill it. `cleanupTimeout` runs its code uninterruptiblyMasked which kind of allows for the code both in `cleanupTimeout` and `handleTimeout` to solve the race with `cleanupTimeout` having a slight advantage due to be uninterruptible.

So in my code, at this point in the code, it means that the running thread finished, but we need to be careful to a potential race so we check the lock. If the lock is `Locked` it means there's already an assassin thread out there so we need to shield ourselves with an `uninterruptibleMask` and kill it. If the lock is `NotLocked` it means the timer hasn't expired and we acquire it immediately (notice that our code is sequential so not possible that the state the lock is being changed) and we just remove the timer from the queue and carry on.

So given this my code is wrong, but I just have to run my thread in an uninterruptible mask in order for it to be as faithful as GHC's implementation as possible